Index: src/predict/online/predictors/Predictor.hpp
===================================================================
--- src/predict/online/predictors/Predictor.hpp	(revision 56548)
+++ src/predict/online/predictors/Predictor.hpp	(working copy)
@@ -54,8 +54,6 @@
             uint32_t vurl_id, entity_id_t publisher_id,
             rm_string *detail = NULL) const;
 
-    uint64_t getPubclass(uint64_t pearl_id) const;
-
     const char* getCacheVersion() const;
 
     double getClickModifier(predict_key& key, rm_string *detail) const;
@@ -141,15 +139,19 @@
     
     void resetThrottleInt();
 
+    uint64_t getPubclass(uint64_t pearl_id) const;
+
     protected:
 
     bool fIsValid;
     PredictCache *fCache;
     PredictorSpec fSpec;
-    mutable PredictRWLock fCacheLock;
-    mutable PredictRWLock fThrottleLock;
     PredictConf fConf;
     vbt_config fVbt;
+    mutable PredictRWLock fThrottleLock;
+
+    public:
+    mutable PredictRWLock fCacheLock;
 };
 
 #include "Predictor.ipp"
